File: project_euler/multiple_of_3_or_5.ts
// https://projecteuler.net/problem=1
// https://math.stackexchange.com/a/9305
const numbers = [...Array(999)].map((_, index) => index + 1);

const multiplesOfThreeOrFive = (numbers: number[], memo: number[] = []): number[] => {
	if (!numbers.length) return memo;
	const current = numbers.shift() as number;
	if (!memo.includes(current) && (!(current % 3) || !(current % 5))) {
		return multiplesOfThreeOrFive(numbers, [current, ...memo]);
	}
	return multiplesOfThreeOrFive(numbers, memo);
}

const allMultiples = multiplesOfThreeOrFive(numbers);
const result = allMultiples.reduce((acc: number, curr: number) => acc + curr, 0);
console.log(result);

File: project_euler/multiple_of_3_or_5.dart
// https://projecteuler.net/problem=1
// https://math.stackexchange.com/a/9305
void main() {
	final growableList = List<int>.generate(999, (int index) => index + 1);
	final List<int> memo = [];
	final allMultiples = multiplesOfThreeOrFive(growableList, memo);
	int sum = 0;
	for (var num in allMultiples) {
		sum += num;
	}
	print(sum);
}

List<int> multiplesOfThreeOrFive(List<int> remainingNumbers, List<int> memo) {
	if (remainingNumbers.length == 0) return memo;
	final lastElement = remainingNumbers.removeLast();
	if (!memo.contains(lastElement) && (lastElement % 3 == 0 || lastElement % 5 == 0)) {
		memo.add(lastElement);
	}
	return multiplesOfThreeOrFive(remainingNumbers, memo);
}